### PR TITLE
[WIP] Experiment: stabilize never type

### PR DESCRIPTION
**Pull request by nikomatsakis**
Follow-up to https://github.com/rust-lang/rust/pull/79366 for the purposes of running crater

### CODE REVIEW

File: compiler/rustc_codegen_cranelift/example/mini_core_hello_world.rs [hunk 1]
```diff
@@ -402,44 +397,267 @@ pub enum E1 {
...
+    _F8(X),
+    _F9(X),
+    _FA(X),
+    _FB(X),
+    _FC(X),
+    _FD(X),
+    _FE(X),
+    _FF(X),
 
     V3,
```
Comment (gilescope) [inline compiler/rustc_codegen_cranelift/example/mini_core_hello_world.rs:656]: Should we be fixing rustfmt here or telling rustfmt to skip this?

Comment (oli-obk) [reply compiler/rustc_codegen_cranelift/example/mini_core_hello_world.rs:656]: I think you'll want to post any comments on https://github.com/rust-lang/rust/pull/79366 as this PR is just for the purpose of running crater

**Comment by oli-obk** [comment]
I would start crater, but I don't know if just a check suffices or if we also need to run all tests

**Comment by nikomatsakis** [comment]
@craterbot build-and-test

check would probably suffice but it *is* possible for these changes to change the semantics of running code and so forth.

**Comment by nikomatsakis** [comment]
@craterbot run mode=build-and-test

**Comment by matthewjasper** [comment]
Should this also change `Infallible` to `!`?

**Comment by nikomatsakis** [comment]
Oh, good catch.

**Comment by nikomatsakis** [comment]
@craterbot abort

**Comment by nikomatsakis** [comment]
@craterbot check

I'm going to start with a check build and we can consider a run-and-test check later.

**Comment by nikomatsakis** [comment]
Seems like it'll be a while :)

**Comment by nikomatsakis** [comment]
Relatively few root failures, but definitely some impact. I did a bit of digging but will have to read in more carefully.

One sample error (chmlib v1.0.0):

```
[INFO] [stdout] error[E0277]: the trait bound `Continuation: From<!>` is not satisfied
[INFO] [stdout]    --> src/lib.rs:485:13
[INFO] [stdout]     |
[INFO] [stdout] 485 |         chm.for_each(Filter::all(), |_, _| panic!("Oops..."))
[INFO] [stdout]     |             ^^^^^^^^ the trait `From<!>` is not implemented for `Continuation`
[INFO] [stdout]     |
[INFO] [stdout]     = help: the following implementations were found:
[INFO] [stdout]               <Continuation as From<()>>
[INFO] [stdout]               <Continuation as From<std::result::Result<(), E>>>
[INFO] [stdout]     = note: required because of the requirements on the impl of `Into<Continuation>` for `!`
```

**Comment by tema3210** [comment]
AFAIK there are three root failures: "trait is not implemented for `!`", "never type was propageted as a generic param to smth." and type inference failures.

* The first point was discussed before
* Second is the major problem: tainting generic parameters. For example: [here](https://crater-reports.s3.amazonaws.com/pr-79470/try%23b2f80f6008cdafc35b4472bd46afa16cc6b0cd63/gh/AbsoluteVirtueXI.thread-channel/log.txt) the anonymous future type got a never return type (and caused error). The point is that these error are caused by proper usage of never type: regressed futures used a `loop`, which **is** an expression of type never. I guess to fix this we have either to invent some crazy schema or to break part of a world in the next edition.
* An inference failure only happened [here](https://crater-reports.s3.amazonaws.com/pr-79470/try%23b2f80f6008cdafc35b4472bd46afa16cc6b0cd63/reg/tortuga-0.1.2/log.txt).

**Pull request closed without merging**